pkg/standalone-utils/contracts/relayer/TetuWrapping.sol

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/standalone-utils/ITetuSmartVault.sol";
+import "@balancer-labs/v2-interfaces/contracts/vault/IVault.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Address.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/math/FixedPoint.sol";
+import "./IBaseRelayerLibrary.sol";
+import "./special/TetuShareValueHelper.sol";
+
+/**
+ * @title TetuWrapping
+ * @notice Allows users to wrap and unwrap Tetu tokens
+ * @dev All functions must be payable so they can be called from a multicall involving ETH
+ */
+abstract contract TetuWrapping is IBaseRelayerLibrary {
+    using Address for address payable;
+    using SafeERC20 for IERC20;
+    using FixedPoint for uint256;
+
+    function wrapTetu(
+        ITetuSmartVault wrappedToken,
+        address sender,
+        address recipient,
+        uint256 amount,
+        uint256 outputReference
+    ) external payable {
+        if (_isChainedReference(amount)) {
+            amount = _getChainedReferenceValue(amount);
+        }
+        
+        IERC20 underlying = IERC20(wrappedToken.underlying());
+
+        // The wrap caller is the implicit sender of tokens, so if the goal is for the tokens
+        // to be sourced from outside the relayer, we must first pull them here.
+        if (sender != address(this)) {
+            require(sender == msg.sender, "Incorrect sender");
+            _pullToken(sender, underlying, amount);
+        }
+        
+        underlying.safeApprove(address(wrappedToken), amount);
+        wrappedToken.deposit(amount);
+        uint256 receivedWrappedAmount = wrappedToken.balanceOf(address(this));
+
+        if (recipient != address(this)) {
+            IERC20(address(wrappedToken)).safeTransfer(recipient, receivedWrappedAmount);
```

Comment:
Although ITetuSmartVault is IERC20, SafeERC20 is not applied to the inherited class. The compiler complained, I needed to explicitly cast

Reply:
Ah, let's change it like this then. Since `ITetuSmartVault is IERC20`, we don't need the intermediate cast to `address`.
```suggestion
            IERC20(wrappedToken).safeTransfer(recipient, receivedWrappedAmount);
```